opentracing_shim/src/main/java/io/opentelemetry/opentracingshim/SpanContextShimTable.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.opentracingshim;
+
+import io.opentelemetry.trace.Span;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nullable;
+
+/*
+ * SpanContextShimTable stores and manages OpenTracing SpanContext instances,
+ * which are expected to a unmodfiable union of SpanContext and Baggage
+ * (DistributedContext/TagMap under OpenTelemetry).
+ *
+ * This requires that changes on a given Span and its (new) SpanContext
+ * are visible in all threads at *any* moment. The current approach uses
+ * a weak map synchronized through a read-write lock to get and set both
+ * SpanContext and any Baggage content related to its owner Span.
+ *
+ * Observe that, because of this design, a global read or write lock
+ * will be taken for ALL operations involving OT SpanContext/Baggage.
+ * When/if performance becomes an issue in the OT Shim layer, consider
+ * adding an extra slot in io.opentelemetry.trace.Span, so:
+ * 1) The current SpanContextShim is directly stored in Span.
+ * 2) The lock for this operation can be on a per-Span basis.
+ *
+ * For more information, see:
+ * https://github.com/opentracing/specification/blob/master/specification.md#set-a-baggage-item
+ */
+final class SpanContextShimTable {
+  private final Map<Span, SpanContextShim> shimsMap = new WeakHashMap<>();
+  private final ReadWriteLock lock = new ReentrantReadWriteLock();
```

Comment:
If all access to the shimsMap needs to be synchronized, why not just wrap it in a synchonized map?

Reply:
In theory, we will get **far more** read accesses than write ones.

Reply:
that's totally fair! thanks!